### PR TITLE
Move plasma retry logic into plasma store provider

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Standardizes the behavior in all of the places in the code that we create objects in plasma. Will also be necessary to trigger global GC (see https://github.com/ray-project/ray/pull/7327).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by edoakes** [comment]
Wasn't sure how to add a test for this, but verified locally that it works for both `ray.put()` and task returns.

**Review by ericl — APPROVED** [review]

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -238,3 +238,9 @@ RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
 /// Note: this only takes effect when gcs service is enabled.
 RAY_CONFIG(int64_t, gcs_service_connect_retries, 50)
 RAY_CONFIG(int64_t, gcs_service_connect_wait_milliseconds, 100)
+
+/// Maximum number of times to retry putting an object when the plasma store is full.
+RAY_CONFIG(uint32_t, object_store_full_max_retries, 5)
+/// Duration to sleep after failing to put an object in plasma because it is full.
+/// This will be exponentially increased for each retry.
+RAY_CONFIG(uint32_t, object_store_full_initial_delay_ms, 1000)
```
Comment (stephanie-wang) [inline src/ray/common/ray_config_def.h]: Yay :)

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -238,3 +238,9 @@ RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
 /// Note: this only takes effect when gcs service is enabled.
 RAY_CONFIG(int64_t, gcs_service_connect_retries, 50)
 RAY_CONFIG(int64_t, gcs_service_connect_wait_milliseconds, 100)
+
+/// Maximum number of times to retry putting an object when the plasma store is full.
+RAY_CONFIG(uint32_t, object_store_full_max_retries, 5)
```
Comment (stephanie-wang) [inline src/ray/common/ray_config_def.h]: Maybe make this an int to allow setting to -1 for infinity.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc [hunk 3]
```diff
@@ -57,26 +57,44 @@ Status CoreWorkerPlasmaStoreProvider::Create(const std::shared_ptr<Buffer> &meta
...
     if (plasma::IsPlasmaObjectExists(status)) {
       RAY_LOG(WARNING) << "Trying to put an object that already existed in plasma: "
                        << object_id << ".";
       return Status::OK();
+    } else if (plasma::IsPlasmaStoreFull(status)) {
+      if (retries >= RayConfig::instance().object_store_full_max_retries()) {
+        std::ostringstream message;
+        message << "Failed to put object " << object_id << " in object store because it "
+                << "is full. Object size is " << data_size << " bytes.";
+        return Status::ObjectStoreFull(message.str());
```
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: nit, could you refactor this function to only have a single `return` statement?

File: src/ray/core_worker/store_provider/plasma_store_provider.cc [hunk 4]
```diff
@@ -55,46 +55,45 @@ Status CoreWorkerPlasmaStoreProvider::Create(const std::shared_ptr<Buffer> &meta
...
                        << "ms for space to free up...";
         usleep(1000 * delay);
         delay *= 2;
         retries += 1;
+        should_retry = true;
       }
+    } else if (plasma::IsPlasmaObjectExists(plasma_status)) {
+      RAY_LOG(WARNING) << "Trying to put an object that already existed in plasma: "
+                       << object_id << ".";
+      status = Status::OK();
```
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: I think this probably isn't needed.

File: src/ray/common/ray_config_def.h [hunk 5]
```diff
@@ -240,6 +240,7 @@ RAY_CONFIG(int64_t, gcs_service_connect_retries, 50)
 RAY_CONFIG(int64_t, gcs_service_connect_wait_milliseconds, 100)
 
 /// Maximum number of times to retry putting an object when the plasma store is full.
+/// Can be set to -1 to enable unlimited retries.
 RAY_CONFIG(uint32_t, object_store_full_max_retries, 5)
```
Comment (stephanie-wang) [inline src/ray/common/ray_config_def.h]: Doesn't this need to be signed?

**Review by stephanie-wang — APPROVED** [review]
A couple small comments, but looks good!

File: src/ray/core_worker/store_provider/plasma_store_provider.cc (hunk 4 above)
Comment (edoakes) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Agreed but don't want to change it in this PR

Comment (stephanie-wang) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Oh I just meant the `status = ...` line, but yeah I guess the whole block could be removed.

**Pull request merged**